content/en/docs/tasks/extend-kubernetes/socks5-proxy-access-api.md

```diff
@@ -0,0 +1,127 @@
+
+* We've got a client machine from where we're going to create requests to talk to the Kubernetes API
+* The Kubernetes server/API is hosted on a remote server.
+* We leverage SSH to create a secure SOCKS5 tunnel between the local and the remote server on top of
+  which the HTTPS traffic between the client and the Kubernetes API will flow.
+
+{{< mermaid >}}
+graph LR;
+
+  subgraph local[Local client machine]
+  client([client])-- local <br> traffic .->  local_ssh[Local SSH <br> SOCKS5 proxy];
+  end
+  local_ssh[SSH <br>SOCKS5 <br> proxy]-- SSH Tunnel -->sshd
+  
+  subgraph remote[Remote server]
+  sshd[SSH <br> server]-- local traffic -->service1;
+  end
+  client([client])-. proxied HTTPs traffic <br> going through the proxy .->service1[Kubernetes API];
+
+  classDef plain fill:#ddd,stroke:#fff,stroke-width:4px,color:#000;
+  classDef k8s fill:#326ce5,stroke:#fff,stroke-width:4px,color:#fff;
+  classDef cluster fill:#fff,stroke:#bbb,stroke-width:2px,color:#326ce5;
+  class ingress,service1,service2,pod1,pod2,pod3,pod4 k8s;
+  class client plain;
+  class cluster cluster;
+{{</ mermaid >}}
+Figure 1. SOCKS5 tutorial components
+
+## Using ssh to create a SOCKS5 proxy
+
+This command starts a SOCKS5 proxy between your client machine and the remote server
+where the Kubernetes API is listening:
+
+```shell
+ssh -D 8080 -q -N username@kubernetes-jump-box.example
+```
+
+* `-D 8080`: opens a SOCKS proxy on local port :8080.
+* `-q`: quiet mode. Causes most warning and diagnostic messages to be suppressed.
+* `-N`: Do not execute a remote command. Useful for just forwarding ports.
+* `username@kubernetes-jump-box.example`: the remote SSH server where the Kubernetes cluster is running.
+
+## Client configuration
+
+To explore the Kubernetes API you'll first need to instruct your clients to send their queries through
+the SOCKS5 proxy we created earlier.
+
+For command-line tools, set the `https_proxy` environment variable and pass it to commands that you run.
+
+```shell
+export https_proxy=socks5://localhost:8080
+```
+
+When you set the `https_proxy` variable, tools such as `curl` route HTTPS traffic through the proxy
+you configured. For this to work, the tool must support SOCKS5 proxying.
+
+{{< note >}}
+In this example localhost will not be the localhost of the client machine but the localhost of
+the remote server.
+It could also be replaced by a hostname that is known by the remote server.
+{{</ note >}}
+
+```shell
+curl -k -v https://localhost/api
+```
+
+To use the official Kubernetes client `kubectl` with a proxy, set the `proxy-url` element
+for the relevant `cluster` entry within  your `~/.kube/config` file. For example:
+
+```yaml
+apiVersion: v1
+clusters:
+- cluster:
+    certificate-authority-data: LRMEMMW2 # shortened for readability 
+    server: https://localhost            # the "Kubernetes API" in the diagram above
+    proxy-url: socks5://localhost:8080   # the "SSH SOCKS5 proxy" in the diagram above
```

Comment:
I've added a note about where the DNS resolution happens, making the environment variable match what's documented. _However_, I don't know whether `kubectl` and friends resolve hostnames locally or remotely when using SOCKS v5. So this might be wrong and instead we need to change the meaning of the note about name resolution.

If kubectl does default to having name resolution performed by the remote peer, we could add a comment about why the URL scheme here is not `socks5h` (I presume the answer would then be: DNS resolution over socks is the default).

Reply:
I'm 99% sure that resolution is done by the remote peer but I'll need to validate that in a lab. I'll keep this comment open for now.

Reply:
From what I've [read](https://github.com/hashicorp/terraform/issues/17754#issuecomment-386411167) and tested kubectl supports remote DNS resolution through socks5 proxies without `socks5h://` or `socks5://` in front of the proxy URL.